### DISCUSSION TITLE
How to merge "tilezed" image into one image

### DISCUSSION THREAD

**Discussion by jurijsk** [qpdf/qpdf#602] [Q&A] [answered]
Hello community!

I'm new to PFDs, and only trying to understand PDF and its metaphysics. 

I have a PDF with illustration which are all split into little tiles, and I would like to merge related tiles back into single image.

I wonder if this "tilezation" is a part of PDF specification and or this is just the result some tool, which was used to generate the PDF, internal representation of image.

Can I use qpdf to extract the whole image instead of separate tiles, or transform the pdf using qpdf to understand that this set of tiles belong to single image/object on the page?

Thank you!

**Comment by jberkenbilt** [comment, marked as answer]
Hello! I'll try to address this. The punchline is that the tiling of the image would be about the tool that created the PDF and is not in any way part of PDF itself. qpdf doesn't have any capability to reassemble the image. Your best bet would be to use a tool like ghostscript to render the page as an image. Tools like ImageMagick and GraphicsMagick can convert PDF files into images, though they are basically using ghostscript under the covers.

To go one level deeper, PDF is basically just a "language" (not a full programming language like PostScript though) for describing the layout of the page. Things like images can get onto the page in a variety of ways to the extent that writing a generalized tool to reassemble the image would be quite challenging. When a raster image is included in PDF, what is essentially happening is that there is an instruction to take a particular raster image and fit it into a rectangular region of the page. That rectangular region can in turn be transformed in various ways to scale, skew, rotate, or perform any other linear transformation on it. Vector graphics, raster images, and text can be freely mixed.

There are some image formats, notably TIFF, which originated from a company that was later acquired by Adobe, internally represent images in either strips or tiles, and some software that converts TIFF images to PDF images converts each strip or tile to a separate raster within PDF. This was a pretty common practice especially when memory was more scarce or when representing images generated by devices like scanners that tend to send the image back in pieces.

I hope this helps. This is a good question. I'm going to tag this as something I might like to add to a FAQ or wiki.

**Reply by jurijsk** [reply]
@jberkenbilt Thank you for eloquent answer. Analogy to other programming languages is really good and illustrates that are multiple ways to do similar things. And this is what I seen converting test files to `svg` and `html` using `mutool`. I'm planning to employ `mutool` to convert the page to `png` and crop out the image part. But from what I can see, defining bounding box of the image isn't a trivial task either?

**Reply by jberkenbilt** [reply]
No, unfortunately, it's not at all trivial in the general case. There are many factors and many places where transformations can happen. The page itself can have transformations applied to it. The image can be rendered in a content stream that can have any number of transformation matrices applied. Or it could be rendered in a form XObject with its own transformations. You don't quite need a full-fledged PDF interpreter to know what the bounding box ends up being, but you have to be able to process content streams, traversing into form XObjects, and understand the transformation matrices, etc. qpdf has some support for doing some of this, but it definitely doesn't provide everything you would need. It's possible that other higher level libraries may do this. I'm not all that knowledgeable about the other libraries. I know pdfjs is pretty capable for javascript, and I know there are several Java libraries, but I've only ever used qpdf!
